### ISSUE TITLE
[Issue] Using Access Token as a bicep parameter

### ISSUE THREAD

**Issue by fangjian0423** [Azure/azure-dev#1211]
hi,

As I know, the variables defined in main.parameters.json file will be replaced by azd, and there are 2 ways to do that:

1. replace by the values of `.azure/{env}/.env` file or environment variables

```json
"param": {
    "value": "${AZURE_ACCESS_TOKEN}"
 }
```

2.  use `secretOrRandomPassword` keyword to get value from KeyVault to replace variables

```json
"param": {
    "value": "$(secretOrRandomPassword ${AZURE_KEY_VAULT_NAME} sqlAdminPassword)"
},
```

Is there any ways to use access token as a bicep parameter?

If I use bicep directly, I can get access token as a parameter, not sure azd can do it or not:

```
az deployment sub create --location centralus --template-file test.bicep --parameters accessToken=$(az account get-access-token --resource-type oss-rdbms --output tsv --query accessToken)
```

Thanks.

**Comment by rajeshkamal5050** [comment]
Don't think azd supports this from the CLI params today @jongio @ellismg?

**Comment by weikanglim** [comment]
@fangjian0423 You're right that the types of expansion `azd` does on the parameters file would be:
1. Environment variable/ env file substitution
2. `secretOrRandomPassword` for keyvault

https://github.com/Azure/azure-dev/blob/446c00c238674760103edcf0b050fa610f052131/cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go#L712-L749

What is the scenario here? Would love to understand more.

**Comment by fangjian0423** [comment]
Sorry for late reply.

> What is the scenario here? Would love to understand more.

I want to contribute postgresql passwordless AZD template. By reading this [docs](https://learn.microsoft.com/en-us/azure/developer/java/spring-framework/migrate-postgresql-to-passwordless-connection?tabs=sign-in-azure-cli%2Cjava%2Caks%2Cassign-role-azure-cli#22-create-a-postgresql-non-admin-user-and-grant-permission), it needs `Access Token` as password to login postgresql:

```bash
psql "host=$AZ_DATABASE_SERVER_NAME.postgres.database.azure.com user=$CURRENT_USERNAME dbname=postgres port=5432 password=$(az account get-access-token --resource-type oss-rdbms --output tsv --query accessToken) sslmode=require" < create_ad_user_mi.sql
```

So I need `Access Token` as a parameter to login.

BTW, I use terraform to achieve this and create 2 issues: https://github.com/Azure/awesome-azd/issues/76 and https://github.com/Azure/awesome-azd/issues/77. Could you pls take a look? thanks.

**Comment by jongio** [comment]
My concern with this approach is the access token and being unsure what risk we are introducing by putting that in a parameter.  Can we figure out a way to do this without using the access token as a param?

**Comment by fangjian0423** [comment]
> My concern with this approach is the access token and being unsure what risk we are introducing by putting that in a parameter.

It's really a good question. Seems we don't have a good solution to resolve it now.

>  Can we figure out a way to do this without using the access token as a param?

Terraform won't put access token in a parameter. It use `provisioner "local-exec"` to run script in local, we could get rid of risk. So I contribute that 2 templates which infra was provided by terraform.

**Comment by jongio** [comment]
@weikanglim - Thoughts on how we can do this without needing to have access token as a param?

**Comment by vhvb1989** [comment]
IIRC, @ellismg recently updated the way azd creates and parses the ARM parameters file.
Since we are loading the Json into memory during runtime (before doing the deploy request), I think we should be able to replace the value in memory directly.  Azd can define a placeholder to support access token and use it from `main.parameters.json` (at the infra folder level), or even by using a prefix for the name of the parameter directly within the bicep files.

**Comment by jongio** [comment]
I think we can figure out a better arch to not require this as a param at all

**Comment by weikanglim** [comment]
@fangjian0423 It looks like you're intending to use that access token to pass to a SQL client CLI and execute the SQL script on the database. We're working on adding script hooks, that would allow you to run scripts locally with your shell (pwsh / bash) post-deploy or post-provision. Would that solve your problem?

@jongio Speaking generally for sensitive parameters. Currently for `bicep`, we support parameter passing either via `envsubst` in the parameters.json, or via prompting for any unset values. If extremely worried about storing secrets in env variables, currently the prompting via unset could work to avoid that. Alternatively, a product improvement would be to add support for CLI parameters flags passing via `-param` or something similar. 

Note that both `bicep` and `terraform` both support sensitive value decorators that would prevent the value from showing up in logs. The question is whether we also worry about the in-memory process environment variable storage. For access tokens, I can see that passing as a CLI flag with an inline expansion to be most ideal.

**Comment by fangjian0423** [comment]
@weikanglim 

> It looks like you're intending to use that access token to pass to a SQL client CLI and execute the SQL script on the database.

Exactly.

> We're working on adding script hooks, that would allow you to run scripts locally with your shell (pwsh / bash) post-deploy or post-provision. Would that solve your problem?

Sounds good. I have another question: would the script hooks have the parameters which I wanted?

Here is a example for terraform template:

The parameters I wanted could passed [here](https://github.com/fangjian0423/todo-java-postgresql-aca-terraform/blob/main/infra/main.tf#L153), then used by [script](https://github.com/fangjian0423/todo-java-postgresql-aca-terraform/blob/main/infra/modules/passwordless/passwordless.tf#L19).

**Comment by jongio** [comment]
Question for you, it looks like only Azure AD admins can take this approach.  We've found that customers don't have access at that level.  I think we can provide a sample that does that, but thoughts on providing an alternative that doesn't require it? https://learn.microsoft.com/en-us/azure/developer/java/spring-framework/migrate-postgresql-to-passwordless-connection?tabs=sign-in-azure-cli%2Cjava%2Capp-service%2Cassign-role-service-connector#11-enable-azure-ad-based-authentication

**Comment by jongio** [comment]
For the design of this solution, we have a couple of options:

1. Implement it as a script that is called during post provision script hook - that feature isn't yet implemented, but coming soon.
2. Ask user to run the script themselves.
3. Update azd to include a method that gets the access token, like it does a secret today.
4. Add the access token to azd .env using the `azd env set` command - and use that in the param list. 

#1 will run every time provision is run, so you may not want to do that or have checks in place to ensure it doesn't run more than once.

Any method that stores tokens or secrets in .env files should run the risk of the user committing that to source.

**Comment by jongio** [comment]
@fangjian0423 hooks are now available, can you please see if you can use them to solve this?  https://learn.microsoft.com/en-us/azure/developer/azure-developer-cli/azd-extensibility

**Comment by weikanglim** [comment]
Closing due to staleness. Hooks does seem to fit the use case here
